crates/containerd-shim-wasm/src/sandbox/libcontainer_instance.rs

```diff
@@ -0,0 +1,187 @@
+};
+
+use crate::sandbox::{
+    instance::ExitCode,
+    instance_utils::{get_instance_root, instance_exists},
+};
+
+use crate::sandbox::InstanceConfig;
+use std::{path::PathBuf, thread};
+
+use super::{error::Error, instance::Wait, Instance};
+
+/// LibcontainerInstance is a trait that gets implemented by a WASI runtime that
+/// uses youki's libcontainer library as the container runtime.
+/// It provides default implementations for some of the Instance trait methods.
+/// The implementor of this trait is expected to implement the
+/// * `new_youki()`
+/// * `get_exit_code()`
+/// * `get_id()`
+/// * `get_root_dir()`
+/// * `build_container()`
+/// methods.
+pub trait LibcontainerInstance {
+    /// The WASI engine type
+    type E: Send + Sync + Clone;
+
+    /// Create a new instance
+    fn new_libcontainer(id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self;
+
+    /// Get the exit code of the instance
+    fn get_exit_code(&self) -> ExitCode;
+
+    /// Get the ID of the instance
+    fn get_id(&self) -> String;
+
+    /// Get the root directory of the instance
+    fn get_root_dir(&self) -> Result<PathBuf, Error>;
+
+    /// Build the container
+    fn build_container(&self) -> Result<Container, Error>;
+}
+
+/// Default implementation of the Instance trait for YoukiInstance
+/// This implementation uses the libcontainer library to create and start
+/// the container.
+impl<T: LibcontainerInstance> Instance for T {
+    type E = T::E;
+
+    fn new(id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self {
+        Self::new_libcontainer(id, cfg)
+    }
+
+    /// Start the instance
+    /// The returned value should be a unique ID (such as a PID) for the instance.
+    /// Nothing internally should be using this ID, but it is returned to containerd where a user may want to use it.
+    fn start(&self) -> Result<u32, Error> {
+        let id = self.get_id();
+        log::info!("starting instance: {}", id);
+
+        let mut container = self.build_container()?;
+        let code = self.get_exit_code();
+        let pid = container.pid().context("failed to get pid")?;
+
+        container
+            .start()
+            .map_err(|err| Error::Any(anyhow::anyhow!("failed to start container: {}", err)))?;
+
+        thread::spawn(move || {
+            let (lock, cvar) = &*code;
+
+            let status = match waitid(WaitID::Pid(pid), WaitPidFlag::WEXITED) {
+                Ok(WaitStatus::Exited(_, status)) => status,
+                Ok(WaitStatus::Signaled(_, sig, _)) => sig as i32,
+                Ok(_) => 0,
+                Err(e) => {
+                    if e == Errno::ECHILD {
+                        log::info!("no child process");
+                        0
+                    } else {
+                        panic!("waitpid failed: {}", e);
+                    }
+                }
+            } as u32;
+            let mut ec = lock.lock().unwrap();
```

Comment:
```suggestion
            let mut ec = lock.lock().map_err(|err| Error::Any(anyhow::anyhow!("failed get a lockr: {}", err)))?;
```

Reply:
I think its good to let this panic anyway.
If we can't lock something is very wrong (dirty mutex state) and this should crash the program.